enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,963 @@
+    The original OAS behaviors are modified to operate on v1 instead of internal types and moved it into to library-go. None of the spec.host generation/authz, validation,
+    and field defaulting in OAS happen during admission, and that's still the case. They just have shims to convert internal->V1, call the shared library behavior, and
+    convert back V1->internal. In KAS, we don't have the same extension points since we're using the same CR handler that all CRs use. So in KAS, these same behaviors are performed by
+    admission plugins.
+    https://speakerdeck.com/sttts/kubernetes-api-codebase-tour?slide=19 does a good job illustrating why KAS route admission plugins are inert on OCP, but functional for MicroShift.
+    KAS admission plugins aren't invoked on OCP. On microshift, the kube-aggregator dispatches route requests to the in-process apiserver.
+    KAS admission route plugins are inert on OCP because routes never pass through the KAS admission chain there. They're proxied to OAS before that stage.
+    On MicroShift, that's not true and KAS admits+handles the route locally.
+    It was considered to be a good idea to take a similar course i.e. adding a new package under https://github.com/openshift/library-go/tree/master/pkg/route and
+    then adding a new admission plugin in openshift-apiserver for OCP and a new admission plugin for MicroShift, both using the library-go code with new api behaviors.
+    So, shared logic in library-go written against route/v1 types, KAS plugin converting unstructured->routev1, and OAS plugin converting internal->v1.
+    Even if something can be represented by a crd schema, sharing the implementation this way was going to remove the opportunity to introduce subtle differences between OAS-routes and CRD-routes.
+    So, in short it was earlier decided to use add validations(Eg: for an immutable list of prohibited values or combinations of values eg: HSTS) in library-go or admissions(Eg: for a dynamic list of prohibited values where
+    validation of one resource depends on data from another resource) using admission plugins in openshift/openshift-apiserver and openshift/kubernetes for OCP and Microshift respectively.
+    So this would avoid duplication for performing validations using library-go in 2 different places for the same route CR. However, there still will be duplication as we would need same consistent
+    admission plugins if needed to be written in openshift/kubernetes for Microshift and openshift/openshift-apiserver for OCP.
+    However later it was decided from a discussion with API team that route CR validations for OCP to be made in library-go code and for Microshift in openshift/api which should be consistent and well-formed.
+    The reason was just using CEL validations on CRD it is possible to express a wide range of previously impossible rules. The CRD schema provides help to client-side tools and interoperates with future schema checks for compatibility in progress.
```

Comment:
@candita, @miheer, is this clearer?
```suggestion
- Why do we need the kubebuilder validation tags on the Route CRD in openshift/api in addition to validation in library-go?
   Answer:
    The library-go validation is used for server-side validation whereas the OpenAPI schema on the Route CRD may be used for client-side validation.  
    On OCP, kube-apiserver delegates the Route API to openshift-apiserver, which employs the library-go validation to validate Route objects (see <https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go>).  
    MicroShift does not use openshift-apiserver; instead, on MicroShift, kube-apiserver handles the Route API using the Route CRD.  However, kube-apiserver has an admission plugin for the Route CRD that employs the same library-go validation that openshift-apiserver uses (see <https://github.com/openshift/kubernetes/blob/15f19ea2dd700767e5337502aec753d2a6e26905/openshift-kube-apiserver/admission/customresourcevalidation/route/validate_route.go>).
    Defining validation for the same API in both library-go and the OpenAPI schema introduces the risk that someone could inadvertently introduce discrepancies between the library-go validation and the OpenAPI schema, leading to buggy and potentially confusing behavior.  Care must be taken to avoid introducing such discrepancies.  
    Historical note: At one time, MicroShift relied on the OpenAPI schema for server-side validation of the Route CRD, and for this reason, kubebuilder validation tags were added to the Route CRD (see <https://github.com/openshift/api/pull/1228>).  Since then, an admission plugin has been added to kube-apiserver that validates the Route CRD using the same library-go validation on MicroShift that openshift-apiserver uses on OpenShift (see <https://github.com/openshift/kubernetes/pull/1416>).  As a result, the kubebuilder validation tags on the Route CRD are no longer required specifically for MicroShift; however, the kubebuilder validation tags are still required for client-side validation.  
```

Reply:
looks good. I made the changes